Feature/implement recipe viewmodel

2Fick opened the pull request:
# PULL REQUEST: Update Recipe Model and Repository Implementation

## Description
- [x] **What has been changed?**
  - **Added `RecipeRepository` Interface**: Introduced the `RecipeRepository` interface for recipe data operations, including a `random` function that retrieves a list of recipes.
  - **Updated `Recipe` Model**: 
    - Changed the type of `idMeal` from `Long` to `String` to accommodate various identifier formats.
    - Combined `ingredients` and `measurements` into a new field `ingredientsAndMeasurements`, represented as `List<Pair<String, String>>` for improved structure.

- [x] **Why was this change made?**
  - These changes enhance the ability to handle recipe data more flexibly and simplify the management of ingredient and measurement pairs within the app. The repository interface provides a clear contract for fetching recipes, supporting a recipe discovery experience for users.

**Code Snippets**:

1. **Recipe Model**:
   ```kotlin
   data class Recipe(
       val idMeal: String, // Changed from Long to String
       val strMeal: String,
       val strCategory: String?,
       val strArea: String?,
       val strInstructions: String,
       val strMealThumbUrl: String,
       val ingredientsAndMeasurements: List<Pair<String, String>> // Combined field as a list of pairs
   )
   ```

2. **RecipeRepository Interface**:
   ```kotlin
   interface RecipeRepository {
       fun random(nbOfElements: Long, onSuccess: (List<Recipe>) -> Unit, onFailure: (Exception) -> Unit)
   }
   ```

## Checklist
- [ ] Tests added.
- [ ] Documentation updated.
- [x] All CI checks passed.
- [ ] Linked issue/feature (if applicable): #<issue-number>

CapraSnitzel commented on app/src/main/java/com/android/sample/model/recipe/RecipeRepository.kt:
```diff
@@ -0,0 +1,6 @@
+
+interface RecipeRepository {
+
+  fun random(nbOfElements: Long, onSuccess: (List<Recipe>) -> Unit, onFailure: (Exception) -> Unit)
```
You should change the type of the nbOfElements parameter from Long to Int because it is not necessary, an Int is enough for this project.

CapraSnitzel replied:
Like for the data class Recipe, I think it should be helpful to have a description for each parameter as well as having a description of the function in order to know what it does.

CapraSnitzel commented on app/src/main/java/com/android/sample/model/recipe/RecipeRepository.kt:
```diff
@@ -0,0 +1,6 @@
+package com.android.sample.model.recipe
+
+interface RecipeRepository {
```
By looking at the SonarCloud, I saw that there is a suggestion on how to define this interface with only only one function. I let you take a look at it and maybe change it.

2Fick replied:
Regarding the suggestion to make the interface functional, I opted not to turn it into a functional interface because we anticipate adding more functions in the future. This will provide flexibility for expanding the functionality of the repository.

CapraSnitzel replied:
In that case, it's perfect. Thank you for the clarification.

CapraSnitzel commented on app/src/main/java/com/android/sample/model/recipe/Recipe.kt:
```diff
@@ -1,12 +1,11 @@
 package com.android.sample.model.recipe
 
 data class Recipe(
```
Even though the attributes of the data class are explicit, I think it should be helpful for the others to have a documentation of the data class where you explain what each attribute represents

CapraSnitzel reviewed (changes requested):
Overall, everything is correct except for the little problems that I have brought forward which are not major.
But, even though they are not major problems, I will still ask you to make these changes before merging.

2Fick commented:
## Changes Made

- **Updated RecipeRepository**:
    - Added KDoc documentation to clarify the purpose of the interface and its parameters.
  - Changed the type of the `nbOfElements` parameter from `Long` to `Int`.
  ```kotlin
  interface RecipeRepository {
      fun random(nbOfElements: Int, onSuccess: (List<Recipe>) -> Unit, onFailure: (Exception) -> Unit)
  }```

- **Updated RecipesViewModel**:
   - Added KDoc documentation for the class and its methods.
  - Modified the `fetchRandomRecipes` method to accept `Int` instead of `Long`.
  ```kotlin 
    fun fetchRandomRecipes(numberOfRecipes: Int) {
         _loading.value = true
         repository.random(nbOfElements = numberOfRecipes, onSuccess = { randomRecipes ->
             _recipes.value = randomRecipes
             _loading.value = false
         }, onFailure = { exception ->
             _loading.value = false
      })
    } 
  

- **Updated RecipesViewModelTest**:
  - Adjusted the test methods to reflect the changes in parameter types for `fetchRandomRecipes`.
  - Used dummy recipes to simulate repository behavior during tests.

- **Updated Recipe Data Class**:
  - Combined ingredients and measurements into a single field represented as `List<Pair<String, String>>`.
  - Added KDoc documentation for each attribute to explain its purpose.

- **Documentation Improvements**:
  - Added detailed comments throughout the code to enhance clarity and maintainability.

This update addresses the feedback received during the code review and improves the overall code quality.

CapraSnitzel reviewed (approved):
Thank you for making the mentioned modifications and explained the ones that did not need any modifications.
